# 📊 **Dataset**
https://www.kaggle.com/datasets/fahd09/riyadh-restaurants-20k/data



*   19361 entry
*   11 columns (name, categories, address, latitude, longtitude, price, likes, photos, tips,  rating, ratingSignals)



# **Import Libraries📚**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly
import plotly.express as px
import math
import os
import kagglehub

import warnings
warnings.filterwarnings('ignore')

# **Read the data 👓**

In [2]:
# Download latest version
path = kagglehub.dataset_download("fahd09/riyadh-restaurants-20k")

print("Path to dataset files:", path)

df_path = os.path.join(path, 'riyadh_resturants_clean.csv')
df = pd.read_csv(df_path)

df.head(-5)

Using Colab cache for faster access to the 'riyadh-restaurants-20k' dataset.
Path to dataset files: /kaggle/input/riyadh-restaurants-20k


name          categories  \
0                         مطعم وقت الشواية   Afghan Restaurant   
1                             ديوانية عكاظ                Café   
2                      شاهي جمر راعي الجمس         Coffee Shop   
3                               غدير الشام   Afghan Restaurant   
4                                  Dunkin'          Donut Shop   
...                                    ...                 ...   
19351                              Retazza  Italian Restaurant   
19352  JAVA CAFE | Terminal 5 (جافا كافيه)                Café   
19353                            Java Time         Coffee Shop   
19354   Upper Crust- KKIA “Riyadh Airport”                Café   
19355         Fresh Baguettes Tasty Coffee                Café   

                                                 address        lat  \
0                 الرياض 14723, المملكة العربية السعودية  24.518983   
1                 الرياض 14726, المملكة العربية السعودية  24.518511   
2                 الرياض 14726, المملكة العربية السعودية  24.519314   
3                               المملكة العربية السعودية  24.519520   
4                       الرياض, المملكة العربية السعودية  24.525001   
...                                                  ...        ...   
19351    Airport, الرياض 13455, المملكة العربية السعودية  24.941585   
19352                           المملكة العربية السعودية  24.942412   
19353  King Khalid International Airport (RUH), الريا...  24.942480   
19354  King Khalid International Airport, الرياض, الم...  24.942339   
19355                   الرياض, المملكة العربية السعودية  24.942736   

             lng     price  likes  photos  tips  rating  ratingSignals  
0      46.656981  Moderate    0.0       0     0     NaN            NaN  
1      46.669149     Cheap    0.0       2     0     NaN            NaN  
2      46.670041     Cheap    0.0       0     0     NaN            NaN  
3      46.671660  Moderate    0.0       0     0     NaN            NaN  
4      46.433944     Cheap   29.0      90     1     8.9           32.0  
...          ...       ...    ...     ...   ...     ...            ...  
19351  46.713071  Moderate    3.0      10     1     NaN            NaN  
19352  46.713373     Cheap    4.0       8     0     NaN            NaN  
19353  46.712140     Cheap    5.0       6     2     7.5            8.0  
19354  46.712199     Cheap    7.0      15     8     5.1           19.0  
19355  46.711950     Cheap    0.0       4     0     NaN            NaN  

[19356 rows x 11 columns]

# **EDA📃**

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19361 entries, 0 to 19360
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           19361 non-null  object 
 1   categories     19361 non-null  object 
 2   address        19361 non-null  object 
 3   lat            19361 non-null  float64
 4   lng            19361 non-null  float64
 5   price          17846 non-null  object 
 6   likes          19359 non-null  float64
 7   photos         19361 non-null  int64  
 8   tips           19361 non-null  int64  
 9   rating         7949 non-null   float64
 10  ratingSignals  7949 non-null   float64
dtypes: float64(5), int64(2), object(4)
memory usage: 1.6+ MB


>  Dataset has 19361 entries with 10 columns. Mix of data types (int, float, object).

**Check for Missing Values**

In [4]:
def check_missing_values(df):
  missing_values = df.isnull().sum()
  print("Missing Values per Column:")
  print(missing_values[missing_values > 0])
  if missing_values.any():
    print("Handle the missing values")
  else:
    print("\nNo Missing Values Found.")

check_missing_values(df)

Missing Values per Column:
price             1515
likes                2
rating           11412
ratingSignals    11412
dtype: int64
Handle the missing values


In [5]:
# Analyze missing values
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_data = pd.DataFrame({
    'Column': missing_percentage.index,
    'Missing_Percentage': missing_percentage.values
})
missing_data = missing_data[missing_data['Missing_Percentage'] > 0].sort_values('Missing_Percentage', ascending=False)

print("Missing Data Analysis:")
missing_data.head(10)

Missing Data Analysis:


Column  Missing_Percentage
10  ratingSignals           58.943236
9          rating           58.943236
5           price            7.825009
6           likes            0.010330

> `ratingSignals` (58.9% missing), `rating` (58.9% missing), `price` (7.8% missing), `likes` (0.01% missing)

**Check for Duplicate Values & Drop them**

In [6]:
def check_duplicates(df):
  duplicates = df.duplicated().sum()
  print(f"Number of Duplicate Samples: {duplicates}")
  if duplicates > 0:
    print("Dropping Duplicates...")
    df.drop_duplicates(inplace=True)
    print("Duplicates Dropped.")
  else:
    print("No Duplicate Samples Found.")

check_duplicates(df)

Number of Duplicate Samples: 1
Dropping Duplicates...
Duplicates Dropped.


**Summary Statistics**

In [7]:
# Descriptive statistics for numerical columns
df.describe()

lat           lng         likes        photos          tips  \
count  19360.000000  19360.000000  19358.000000  19360.000000  19360.000000   
mean      24.722584     46.697343     63.612305    167.781612     17.091632   
std        0.082220      0.074111    291.035406    985.449662     75.458752   
min       24.469683     46.272777      0.000000      0.000000      0.000000   
25%       24.674720     46.643277      0.000000      1.000000      0.000000   
50%       24.737770     46.693641      3.000000      4.000000      1.000000   
75%       24.784009     46.753452     17.000000     24.000000      5.000000   
max       24.943424     46.940912   9984.000000  26332.000000   2301.000000   

            rating  ratingSignals  
count  7948.000000    7948.000000  
mean      7.536047     201.660544  
std       0.942885     552.256003  
min       4.400000       0.000000  
25%       6.900000      14.000000  
50%       7.600000      38.000000  
75%       8.200000     143.000000  
max       9.600000   12376.000000

---

# **Data Cleaning🧽**

> copy the dataset for cleaning

In [8]:
df_clean = df.copy()

> fill missing values with 'unknown'

In [9]:
df_clean['price'] = df_clean['price'].fillna('unknown')
df_clean['likes'] = df_clean['likes'].fillna(df_clean['likes'].mode()[0])
df_clean['rating'] = df_clean['rating'].fillna(df_clean['rating'].mode()[0])
df_clean['ratingSignals'] =df_clean['ratingSignals'].fillna(df_clean['ratingSignals'].mode()[0])
check_missing_values(df_clean)

Missing Values per Column:
Series([], dtype: int64)

No Missing Values Found.


> drop cafes in the dataset

In [10]:
df_clean = df_clean[~df_clean["categories"].str.lower().str.contains("cafe|coffee|café|donut|dessert|ice cream", na=False)]

In [11]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12358 entries, 0 to 19360
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           12358 non-null  object 
 1   categories     12358 non-null  object 
 2   address        12358 non-null  object 
 3   lat            12358 non-null  float64
 4   lng            12358 non-null  float64
 5   price          12358 non-null  object 
 6   likes          12358 non-null  float64
 7   photos         12358 non-null  int64  
 8   tips           12358 non-null  int64  
 9   rating         12358 non-null  float64
 10  ratingSignals  12358 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 1.1+ MB


# **Plots📊**

In [12]:
p1 = {'lat': 24.719462, 'lon': 46.719108}
df.loc[pd.isnull(df['price']), 'price'] = 'Unknown'
fig = px.scatter_mapbox(df,
                        lat=df['lat'],
                        lon=df['lng'],
                        center=p1,
                        color=df['price'],
                        zoom=9,
                        mapbox_style="open-street-map",
                        title='Riyadh Restaurants'
                       )
fig.show()

In [13]:
p1 = {'lat': 24.719462, 'lon': 46.719108}
fig = px.density_mapbox(
    df_clean,
    lat="lat",
    lon="lng",
    z="likes",
    radius=15,
    center=p1,
    zoom=10,
    mapbox_style="open-street-map",
    title="Popular Spots by Likes"
)

fig.show()


In [15]:
fig = px.scatter_mapbox(
    df_clean,
    lat="lat",
    lon="lng",
    size="likes",
    color="price",
    size_max=25,
    zoom=10,
    mapbox_style="open-street-map",
    hover_name="name",
    hover_data=["categories", "rating", "likes"],
    title="Popularity Map"
)

fig.show()


In [16]:
fig = px.scatter(
    df_clean,
    x="rating",
    y="likes",
    color="price",
    size="likes",
    hover_name="name",
    hover_data=["categories"],
    log_y=True,
    title="Rating vs Popularity by Price"
)

fig.show()


In [18]:
fig = px.pie(
    df_clean,
    names="price",
    hole=0.45,
    title="Restaurant Price Distribution in Riyadh"
)

fig.show()


In [22]:
fig = px.scatter_mapbox(
    df_clean,
    lat="lat",
    lon="lng",
    color="price",
    size="rating",
    animation_frame="rating",
    zoom=10,
    mapbox_style="open-street-map",
    title="Restaurants by Rating"
)

fig.show()
